### ISSUE TITLE
[java] ProperCloneImplementation not valid for final class

### ISSUE THREAD

**Issue by numeralnathan** [pmd/pmd#2410]
**Affects PMD Version:**

6.21.0

**Rule:**

[ProperCloneImplementation](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#propercloneimplementation)

**Description:**

The rule `ProperCloneImplementation` flags a problem on `clone()`.  However, it is my opinion that this is perfectly valid since `MyClass` is declared `final`.  Is my opinion correct or does it need corrected?

The documentation should be enhanced to state that declaring the class as final is sufficient.

**Code Sample demonstrating the issue:**

```java
public final class MyClass implements Cloneable
{
   public MyClass clone()
   {
      return new MyClass();
   }
```

**Expected outcome:**

PMD is showing a false-positive

**Running PMD through:**

Eclipse

**Comment by adangel** [comment]
I think, you are right. This is described e.g. in Effective Java, Item 11 (available e.g. here: http://jtechies.blogspot.com/2012/07/item-11-override-clone-judiciously.html)

The clone method must return a instance of the class being cloned. That means in your example:

``` java
MyClass mc = new MyClass();
Object clone = mc.clone();
assert mc.getClass() == clone.getClass(); // this is true
```

Since MyClass is final and therefore cannot have sub classes, this is fine. The returned type is always correct.

If MyClass would *not* be final then it could have a subclass (e.g. MySubClass). The implementation would be wrong:

> In practice, programmers assume that if they extend a class and invoke super.clone from the subclass, the returned object will be an instance of the subclass. 


```java
class MySubClass extends MyClass implements Cloneable {
    public MySubClass clone() {
        return (MySubClass) super.clone(); // runtime error
    }
}
```

The only way to get the correct class while cloning when being a "intermediate" class (like the non-final MyClass example) is to call super.clone() - which eventually calls Object's clone() - which creates a new instance of the correct subtype. Otherwise the intermediate class would need to know about all the subtypes (which is not possible as long as the class it not sealed).
